pkg/operator/ceph/disruption/clusterdisruption/osd.go

```diff
@@ -168,130 +182,192 @@ func (r *ReconcileClusterDisruption) reconcilePDBsForOSDs(
+	}
+
+	activeDrains := len(drainingFailureDomains) > 0
+
+	// create blocking PDBs for all the OSDs that are not being drained right now
+	if activeDrains {
+		logger.Infof("pg health: %q. detected drains on %q: %v", pgHealthMsg, failureDomain, drainingFailureDomains)
+		err := r.handleActiveDrains(allFailureDomains, drainingFailureDomains, failureDomain, clusterInfo.Namespace, clean)
+		if err != nil {
+			return errors.Wrap(err, "failed to handle active drains")
+		}
 	}
+
 	if pdbStateMap.Data == nil {
 		pdbStateMap.Data = make(map[string]string)
 	}
-	_, ok := pdbStateMap.Data[disabledPDBKey]
+	_, ok := pdbStateMap.Data[drainingFailureDomain]
 	if !ok {
-		pdbStateMap.Data[disabledPDBKey] = ""
+		pdbStateMap.Data[drainingFailureDomain] = ""
 	}
-	var disabledPDBTime time.Time
-	timeString, disabledPDBTimeSet := pdbStateMap.Data[disabledPDBTimeKey]
-	if disabledPDBTimeSet {
-		disabledPDBTime, err = time.Parse(time.RFC3339, timeString)
+	var drainingFDTime time.Time
+	timeString, drainingFDTimeSet := pdbStateMap.Data[drainingFDTimeKey]
+	if drainingFDTimeSet {
+		drainingFDTime, err = time.Parse(time.RFC3339, timeString)
 		if err != nil {
-			logger.Errorf("could not parse timestamp %v. %v", disabledPDBTime, err)
-			disabledPDBTime = time.Now()
-			pdbStateMap.Data[disabledPDBTimeKey] = disabledPDBTime.Format(time.RFC3339)
+			logger.Errorf("could not parse timestamp %v. %v", drainingFDTime, err)
+			drainingFDTime = time.Now()
+			pdbStateMap.Data[drainingFDTimeKey] = drainingFDTime.Format(time.RFC3339)
 		}
 	}
-	recentlyChanged := disabledPDBTimeSet && time.Since(disabledPDBTime) < time.Minute
-	shouldChange := clean && !recentlyChanged
-	activeDrains := len(drainingFailureDomains) != 0
-	if activeDrains {
-		logger.Infof("pg health: %q. detected drains on %q: %v", pgHealthMsg, poolFailureDomain, drainingFailureDomains)
-	}
-	if shouldChange {
-		if activeDrains {
-			pdbStateMap.Data[disabledPDBKey] = drainingFailureDomains[0]
-			pdbStateMap.Data[disabledPDBTimeKey] = time.Now().Format(time.RFC3339)
-		} else {
-			pdbStateMap.Data[disabledPDBKey] = ""
-			delete(pdbStateMap.Data, disabledPDBTimeKey)
+
+	clusterHealed := !activeDrains && clean
+	if clusterHealed {
+		pdbStateMap.Data[drainingFailureDomain] = ""
+		delete(pdbStateMap.Data, drainingFDTimeKey)
+	} else if activeDrains {
+		currentlyDrainingFD, ok := pdbStateMap.Data[drainingFailureDomain]
+		if !ok || drainingFailureDomains[0] != currentlyDrainingFD {
+			pdbStateMap.Data[drainingFailureDomain] = drainingFailureDomains[0]
+			pdbStateMap.Data[drainingFDTimeKey] = time.Now().Format(time.RFC3339)
 		}
 	}
 
-	err = r.updateNoout(clusterInfo, pdbStateMap, allFailureDomainsMap)
+	err = r.updateNoout(clusterInfo, pdbStateMap, allFailureDomains)
 	if err != nil {
-		logger.Errorf("could not update maintenance noout in cluster %q with ceph image. %v", request, err)
+		logger.Errorf("failed to update maintenance noout in cluster %q with ceph image. %v", request, err)
```

Comment:
"with ceph image"? This statement is a bit confusing IMO.

Reply:
indeed. Removed `with ceph image`.